src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs

```diff
@@ -1207,87 +1481,242 @@ private AsnWriter ExportSubjectPublicKeyInfoCore()
         {
-            ThrowIfDisposed();
-
-            // TODO: Determine a more appropriate maximum size once the format is actually known.
-            int size = Algorithm.SecretKeySizeInBytes * 2;
-            // The buffer is only being passed out as a span, so the derived type can't meaningfully
-            // hold on to it without being malicious.
-            byte[] rented = CryptoPool.Rent(size);
-            int written;
-
-            while (!TryExportPkcs8PrivateKey(rented, out written))
+            AsnWriter tmp = ExportPkcs8PrivateKeyCallback(static pkcs8 =>
             {
-                size = rented.Length;
-                CryptoPool.Return(rented, 0);
-                rented = CryptoPool.Rent(size * 2);
-            }
+                AsnWriter writer = new(AsnEncodingRules.BER, initialCapacity: pkcs8.Length);
+                try
+                {
+                    writer.WriteEncodedValueForCrypto(pkcs8);
+                }
+                catch
+                {
+                    writer.Reset();
+                    throw;
+                }
 
-            AsnWriter tmp = new AsnWriter(AsnEncodingRules.BER);
+                return writer;
+            });
 
             try
             {
-                tmp.WriteEncodedValueForCrypto(rented.AsSpan(0, written));
                 return KeyFormatHelper.WriteEncryptedPkcs8(passwordBytes, tmp, pbeParameters);
             }
             finally
             {
                 tmp.Reset();
-                CryptoPool.Return(rented, written);
             }
         }
 
         private AsnWriter ExportEncryptedPkcs8PrivateKeyCore(ReadOnlySpan<char> password, PbeParameters pbeParameters)
         {
-            ThrowIfDisposed();
+            AsnWriter tmp = ExportPkcs8PrivateKeyCallback(static pkcs8 =>
+            {
+                AsnWriter writer = new(AsnEncodingRules.BER, initialCapacity: pkcs8.Length);
+                try
+                {
+                    writer.WriteEncodedValueForCrypto(pkcs8);
+                }
+                catch
+                {
+                    writer.Reset();
+                    throw;
+                }
+
+                return writer;
+            });
 
-            // TODO: Determine a more appropriate maximum size once the format is actually known.
-            int initialSize = Algorithm.SecretKeySizeInBytes * 2;
+            try
+            {
+                return KeyFormatHelper.WriteEncryptedPkcs8(password, tmp, pbeParameters);
+            }
+            finally
+            {
+                tmp.Reset();
+            }
+        }
+
+        private TResult ExportPkcs8PrivateKeyCallback<TResult>(ExportPkcs8PrivateKeyFunc<TResult> func)
+        {
+            // A PKCS#8 ML-DSA secret key has an ASN.1 overhead of 28 bytes, assuming no attributes.
+            // Make it an even 32 and that should give a good starting point for a buffer size.
```

Comment:
Computed with `openssl genpkey -algorithm ML-DSA-87 | openssl pkey -provparam ml-dsa.output_formats=priv-only -outform DER | wc` and compared to secret key size.